### ISSUE TITLE
windows opencl compile error

### ISSUE THREAD

**Issue by isriam** [fireice-uk/xmr-stak#2235]
Please provide as much as possible information to reproduce the issue.

Open CL with OCL_SDK does not compile.  Gives 18 errors related to xmstak_opencl_backend.vcproj

# Basic information
  - Type of the CPU. 
intel i5
  - Type of the GPU (if you try to miner with the GPU).
rx 480

# Compile issues
  - Which OS do you use?
windows 10
  ```
  add **all** commands you used and the **full** compile output here
  ```
cmake -G "Visual Studio 15 2017 Win64" -DCUDA_ENABLE=OFF -DOpenCL_INCLUDE_DIR='C:\Program Files (x86)\OCL_SDK_Light\include' -DOpenCL_LIBRARY='C:\Program Files (x86)\OCL_SDK_Light\lib\x86_64' v141,host=x64 ..

cmake --build . --config Release --target install
  ```
  run `cmake -LA .` in the build folder and add the output here
  ```
CMake Error: The source directory "D:/My Documents/Downloads/xmr-stak-2.8.2/xmr-stak-2.8.2/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
-- Cache values
CMAKE_ASM_MASM_COMPILER:FILEPATH=D:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/ml64.exe
CMAKE_ASM_MASM_FLAGS:STRING=
CMAKE_ASM_MASM_FLAGS_DEBUG:STRING=
CMAKE_ASM_MASM_FLAGS_MINSIZEREL:STRING=
CMAKE_ASM_MASM_FLAGS_RELEASE:STRING=
CMAKE_ASM_MASM_FLAGS_RELWITHDEBINFO:STRING=
CMAKE_BUILD_TYPE:STRING=Release
CMAKE_CONFIGURATION_TYPES:STRING=Debug;Release;MinSizeRel;RelWithDebInfo
CMAKE_CXX_FLAGS:STRING=/DWIN32 /D_WINDOWS /W3 /GR /EHsc
CMAKE_CXX_FLAGS_DEBUG:STRING=/MDd /Zi /Ob0 /Od /RTC1
CMAKE_CXX_FLAGS_MINSIZEREL:STRING=/MD /O1 /Ob1 /DNDEBUG
CMAKE_CXX_FLAGS_RELEASE:STRING=/MD /O2 /Ob2 /DNDEBUG
CMAKE_CXX_FLAGS_RELWITHDEBINFO:STRING=/MD /Zi /O2 /Ob1 /DNDEBUG
CMAKE_CXX_STANDARD_LIBRARIES:STRING=kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib
CMAKE_C_FLAGS:STRING=/DWIN32 /D_WINDOWS /W3
CMAKE_C_FLAGS_DEBUG:STRING=/MDd /Zi /Ob0 /Od /RTC1
CMAKE_C_FLAGS_MINSIZEREL:STRING=/MD /O1 /Ob1 /DNDEBUG
CMAKE_C_FLAGS_RELEASE:STRING=/MD /O2 /Ob2 /DNDEBUG
CMAKE_C_FLAGS_RELWITHDEBINFO:STRING=/MD /Zi /O2 /Ob1 /DNDEBUG
CMAKE_C_STANDARD_LIBRARIES:STRING=kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib
CMAKE_EXE_LINKER_FLAGS:STRING=/machine:x64
CMAKE_EXE_LINKER_FLAGS_DEBUG:STRING=/debug /INCREMENTAL
CMAKE_EXE_LINKER_FLAGS_MINSIZEREL:STRING=/INCREMENTAL:NO
CMAKE_EXE_LINKER_FLAGS_RELEASE:STRING=/INCREMENTAL:NO
CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO:STRING=/debug /INCREMENTAL
CMAKE_INSTALL_PREFIX:PATH=D:/My Documents/Downloads/xmr-stak-2.8.2/xmr-stak-2.8.2/build
CMAKE_LINKER:FILEPATH=D:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/link.exe
CMAKE_LINK_STATIC:BOOL=OFF
CMAKE_MODULE_LINKER_FLAGS:STRING=/machine:x64
CMAKE_MODULE_LINKER_FLAGS_DEBUG:STRING=/debug /INCREMENTAL
CMAKE_MODULE_LINKER_FLAGS_MINSIZEREL:STRING=/INCREMENTAL:NO
CMAKE_MODULE_LINKER_FLAGS_RELEASE:STRING=/INCREMENTAL:NO
CMAKE_MODULE_LINKER_FLAGS_RELWITHDEBINFO:STRING=/debug /INCREMENTAL
CMAKE_RC_COMPILER:FILEPATH=rc
CMAKE_RC_FLAGS:STRING=/DWIN32
CMAKE_RC_FLAGS_DEBUG:STRING=/D_DEBUG
CMAKE_RC_FLAGS_MINSIZEREL:STRING=
CMAKE_RC_FLAGS_RELEASE:STRING=
CMAKE_RC_FLAGS_RELWITHDEBINFO:STRING=
CMAKE_SHARED_LINKER_FLAGS:STRING=/machine:x64
CMAKE_SHARED_LINKER_FLAGS_DEBUG:STRING=/debug /INCREMENTAL
CMAKE_SHARED_LINKER_FLAGS_MINSIZEREL:STRING=/INCREMENTAL:NO
CMAKE_SHARED_LINKER_FLAGS_RELEASE:STRING=/INCREMENTAL:NO
CMAKE_SHARED_LINKER_FLAGS_RELWITHDEBINFO:STRING=/debug /INCREMENTAL
CMAKE_SKIP_INSTALL_RPATH:BOOL=NO
CMAKE_SKIP_RPATH:BOOL=NO
CMAKE_STATIC_LINKER_FLAGS:STRING=/machine:x64
CMAKE_STATIC_LINKER_FLAGS_DEBUG:STRING=
CMAKE_STATIC_LINKER_FLAGS_MINSIZEREL:STRING=
CMAKE_STATIC_LINKER_FLAGS_RELEASE:STRING=
CMAKE_STATIC_LINKER_FLAGS_RELWITHDEBINFO:STRING=
CMAKE_VERBOSE_MAKEFILE:BOOL=FALSE
CPU_ENABLE:BOOL=ON
CUDA_ENABLE:BOOL=OFF
EXECUTABLE_OUTPUT_PATH:STRING=bin
HWLOC:FILEPATH=C:/xmr-stak-dep/hwloc/lib/libhwloc.lib
HWLOC_ENABLE:BOOL=ON
HWLOC_INCLUDE_DIR:PATH=C:/xmr-stak-dep/hwloc/include
LIBRARY_OUTPUT_PATH:STRING=bin
LIB_EAY_DEBUG:FILEPATH=LIB_EAY_DEBUG-NOTFOUND
LIB_EAY_LIBRARY_DEBUG:FILEPATH=LIB_EAY_LIBRARY_DEBUG-NOTFOUND
LIB_EAY_RELEASE:FILEPATH=C:/xmr-stak-dep/openssl/lib/libeay32.lib
MHTD:FILEPATH=C:/xmr-stak-dep/libmicrohttpd/lib/libmicrohttpd.lib
MICROHTTPD_ENABLE:BOOL=ON
MTHD_INCLUDE_DIR:PATH=C:/xmr-stak-dep/libmicrohttpd/include
OPENSSL_INCLUDE_DIR:PATH=C:/xmr-stak-dep/openssl/include
OpenCL_ENABLE:BOOL=ON
OpenCL_INCLUDE_DIR:PATH=C:/Program Files (x86)/OCL_SDK_Light/include
OpenCL_LIBRARY:FILEPATH=C:/Program Files (x86)/OCL_SDK_Light/lib/x86_64
OpenSSL_ENABLE:BOOL=ON
SSL_EAY_DEBUG:FILEPATH=SSL_EAY_DEBUG-NOTFOUND
SSL_EAY_LIBRARY_DEBUG:FILEPATH=SSL_EAY_LIBRARY_DEBUG-NOTFOUND
SSL_EAY_RELEASE:FILEPATH=C:/xmr-stak-dep/openssl/lib/ssleay32.lib
XMR-STAK_COMPILE:STRING=native

# Issue with the execution
  - Do you compiled the miner by our own?
  ```
  run `./xmr-stak --version-long` and add the output here
  ```

# AMD OpenCl issue

  ```
  run `clinfo` and add the output here
  ```
Number of platforms:                             1
  Platform Profile:                              FULL_PROFILE
  Platform Version:                              OpenCL 2.1 AMD-APP (2766.5)
  Platform Name:                                 AMD Accelerated Parallel Processing
  Platform Vendor:                               Advanced Micro Devices, Inc.
  Platform Extensions:                           cl_khr_icd cl_khr_d3d10_sharing cl_khr_d3d11_sharing cl_khr_dx9_media_sharing cl_amd_event_callback cl_amd_offline_devices

  Platform Name:                                 AMD Accelerated Parallel Processing
Number of devices:                               1
  Device Type:                                   CL_DEVICE_TYPE_GPU
  Vendor ID:                                     1002h
  Board name:                                    Radeon (TM) RX 480 Graphics
  Device Topology:                               PCI[ B#1, D#0, F#0 ]
  Max compute units:                             36
  Max work items dimensions:                     3
    Max work items[0]:                           1024
    Max work items[1]:                           1024
    Max work items[2]:                           1024
  Max work group size:                           256
  Preferred vector width char:                   4
  Preferred vector width short:                  2
  Preferred vector width int:                    1
  Preferred vector width long:                   1
  Preferred vector width float:                  1
  Preferred vector width double:                 1
  Native vector width char:                      4
  Native vector width short:                     2
  Native vector width int:                       1
  Native vector width long:                      1
  Native vector width float:                     1
  Native vector width double:                    1
  Max clock frequency:                           1266Mhz
  Address bits:                                  64
  Max memory allocation:                         3422552064
  Image support:                                 Yes
  Max number of images read arguments:           128
  Max number of images write arguments:          64
  Max image 2D width:                            16384
  Max image 2D height:                           16384
  Max image 3D width:                            2048
  Max image 3D height:                           2048
  Max image 3D depth:                            2048
  Max samplers within kernel:                    16
  Max size of kernel argument:                   1024
  Alignment (bits) of base address:              2048
  Minimum alignment (bytes) for any datatype:    128
  Single precision floating point capability
    Denorms:                                     No
    Quiet NaNs:                                  Yes
    Round to nearest even:                       Yes
    Round to zero:                               Yes
    Round to +ve and infinity:                   Yes
    IEEE754-2008 fused multiply-add:             Yes
  Cache type:                                    Read/Write
  Cache line size:                               64
  Cache size:                                    16384
  Global memory size:                            4294967296
  Constant buffer size:                          3422552064
  Max number of constant args:                   8
  Local memory type:                             Scratchpad
  Local memory size:                             32768
  Max pipe arguments:                            16
  Max pipe active reservations:                  16
  Max pipe packet size:                          3422552064
  Max global variable size:                      3080296704
  Max global variable preferred total size:      4294967296
  Max read/write image args:                     64
  Max on device events:                          1024
  Queue on device max size:                      8388608
  Max on device queues:                          1
  Queue on device preferred size:                262144
  SVM capabilities:
    Coarse grain buffer:                         Yes
    Fine grain buffer:                           Yes
    Fine grain system:                           No
    Atomics:                                     No
  Preferred platform atomic alignment:           0
  Preferred global atomic alignment:             0
  Preferred local atomic alignment:              0
  Kernel Preferred work group size multiple:     64
  Error correction support:                      0
  Unified memory for Host and Device:            0
  Profiling timer resolution:                    1
  Device endianess:                              Little
  Available:                                     Yes
  Compiler available:                            Yes
  Execution capabilities:
    Execute OpenCL kernels:                      Yes
    Execute native function:                     No
  Queue on Host properties:
    Out-of-Order:                                No
    Profiling :                                  Yes
  Queue on Device properties:
    Out-of-Order:                                Yes
    Profiling :                                  Yes
  Platform ID:                                   00007FFC61556FD0
  Name:                                          Ellesmere
  Vendor:                                        Advanced Micro Devices, Inc.
  Device OpenCL C version:                       OpenCL C 2.0
  Driver version:                                2766.5
  Profile:                                       FULL_PROFILE
  Version:                                       OpenCL 2.0 AMD-APP (2766.5)
  Extensions:                                    cl_khr_fp64 cl_amd_fp64 cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_fp16 cl_khr_gl_sharing cl_khr_gl_depth_images cl_amd_device_attribute_query cl_amd_vec3 cl_amd_printf cl_amd_media_ops cl_amd_media_ops2 cl_amd_popcnt cl_khr_d3d10_sharing cl_khr_d3d11_sharing cl_khr_dx9_media_sharing cl_khr_image2d_from_buffer cl_khr_spir cl_khr_subgroups cl_khr_gl_event cl_khr_depth_images cl_khr_mipmap_image cl_khr_mipmap_image_writes cl_amd_liquid_flash cl_amd_planar_yuv

# Stability issue
  - Is the CPU or GPU overclocked?
no/no
  - Is the Main memory of the CPU or GPU undervolted?
no

**Comment by isriam** [comment]
here is the errors in cmake config.  it looks like its not linking to the OCL_SDK correctly.  Also it doesn't appear C++ 14.11 is being used.

PS D:\My Documents\Downloads\xmr-stak-2.8.2\xmr-stak-2.8.2\build> cmake -G "Visual Studio 15 2017 Win64" -DCUDA_ENABLE=O
FF -DOpenCL_INCLUDE_DIR='C:\Program Files (x86)\OCL_SDK_Light\include' -DOpenCL_LIBRARY='C:\Program Files (x86)\OCL_SDK_
Light\lib\x86' v141,host=x64 ..
-- The C compiler identification is MSVC 19.16.27027.1
-- The CXX compiler identification is MSVC 19.16.27027.1
-- Check for working C compiler: D:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe
-- Check for working C compiler: D:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: D:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe
-- Check for working CXX compiler: D:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for CL_VERSION_2_0
-- Looking for CL_VERSION_2_0 - found
-- Found OpenCL: C:/Program Files (x86)/OCL_SDK_Light/lib/x86 (found version "2.0")
-- Found OpenSSL: C:/xmr-stak-dep/openssl/lib/libeay32.lib (found version "1.0.2m")
-- The ASM_MASM compiler identification is MSVC
-- Found assembler: D:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/ml64.exe
-- Configuring done
WARNING: Target "xmrstak_opencl_backend" requests linking to directory "C:/Program Files (x86)/OCL_SDK_Light/lib/x86".  Targets may link only to libraries.  CMake is dropping the item.
WARNING: Target "xmrstak_opencl_backend" requests linking to directory "C:/Program Files (x86)/OCL_SDK_Light/lib/x86".  Targets may link only to libraries.  CMake is dropping the item.
WARNING: Target "xmrstak_opencl_backend" requests linking to directory "C:/Program Files (x86)/OCL_SDK_Light/lib/x86".  Targets may link only to libraries.  CMake is dropping the item.
WARNING: Target "xmrstak_opencl_backend" requests linking to directory "C:/Program Files (x86)/OCL_SDK_Light/lib/x86".  Targets may link only to libraries.  CMake is dropping the item.
-- Generating done
-- Build files have been written to: D:/My Documents/Downloads/xmr-stak-2.8.2/xmr-stak-2.8.2/build
PS D:\My Documents\Downloads\xmr-stak-2.8.2\xmr-stak-2.8.2\build>

**Comment by isriam** [comment]
also i cannot use 14.11 c++ tools for some reason even though its installed.

PS D:\My Documents\Downloads\xmr-stak-2.8.2\xmr-stak-2.8.2> E:\VS2017\VC\Auxiliary\Build\vcvarsall.bat x64 -vcvars_ver='
14.11'
**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.5.1
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
[ERROR:vcvars.bat] Toolset directory for version '14.11' was not found.
[ERROR:VsDevCmd.bat] *** VsDevCmd.bat encountered errors. Environment may be incomplete and/or incorrect. ***

**Comment by Spudz76** [comment]
Now that 14.16 is current, you would have to install 14.11 manually to make that work, very similarly to 14.0;  It's down in the MSVS Installer individual package selections...  but I build with 14.16 and it works great.  14.1x will always be binary/link compatible so 14.16 is essentially the newest 14.11 (no real need to use literally 14.11)

Your cmake command looks wrong too...
* `-DOpenCL_LIBRARY=` should point to a literal library (.lib) file, also you've used `x86` 32-bit somehow, and only the dir.  Find the x86_64 actual lib file location and put that instead
* `v141,host=x64` should be prefixed with `-T` such as `-T v141,host=x64` as that is a Target definition.  MSBuild tools may not setup the toolset correctly without `-T` and I'm a little surprised it didn't complain.

**Comment by Spudz76** [comment]
Repaired CMake commands:
```
cmake -G "Visual Studio 15 2017 Win64" -DCUDA_ENABLE=OFF -DOpenCL_INCLUDE_DIR='C:\Program Files (x86)\OCL_SDK_Light\include' -DOpenCL_LIBRARY='C:\Program Files (x86)\OCL_SDK_Light\lib\x86_64\opencl.lib' -T v141,host=x64 ..

cmake --build .. --config Release --target install
```

**Comment by isriam** [comment]
woah thank you!!  i was trying all kinds of ways to get that library to work.

also the windows compile wiki might need updating with those opencl commands for the new ocl_sdk?

**Comment by Spudz76** [comment]
Yes, and also adding the `OCL_ROOT` environment path (the OCL SDK Light installer sets that) for detecting and locating it in preference of other SDKs.  Working on a pullreq to smooth this out.  It's already working pretty well on my test Win7 box (with both AMD-APP-SDK-3.0 and the OCL SDK Light installed to make sure it picks the best available, in case of dirty systems where the old SDK wasn't removed)

**Comment by Spudz76** [comment]
I have no AMD cards on Windows box to test it in actual runtime, but I can test all the compilation parts just fine.  Did your compile work, and run OK?

**Comment by psychocrypt** [comment]
fixed with #2239
